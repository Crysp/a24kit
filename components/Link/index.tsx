import React from 'react';
import {
    StyledLink,
    Wrapper,
    Addon,
} from './styled';

export interface Props {
    children: React.ReactNode
    href: string
    icon: React.ReactNode
    iconPosition: 'before' | 'after'
    pseudo: boolean
    className: string
    onClick(event: React.SyntheticEvent): void
}

export default class Link extends React.PureComponent<Props> {
    static defaultProps = {
        href: '',
        iconPosition: 'before',
        pseudo: false,
        className: '',
        icon: null,
        onClick: undefined,
    };

    onClick = (e: React.SyntheticEvent) => {
        const { href, onClick } = this.props;
        if (href.length === 0) e.preventDefault();
        if (onClick) onClick(e);
    };

    render() {
        const {
            children,
            href,
            icon,
            iconPosition,
            pseudo,
            ...other
        } = this.props;
        return (
            <StyledLink
                href={href}
                pseudo={pseudo}
                {...other}
                onClick={this.onClick}
            >
                <Wrapper pseudo={pseudo}>
                    {icon && iconPosition === 'before' && <Addon>{icon}</Addon>}
                    <span>{children}</span>
                    {icon && iconPosition === 'after' && <Addon>{icon}</Addon>}
                </Wrapper>
            </StyledLink>
        );
    }
}
