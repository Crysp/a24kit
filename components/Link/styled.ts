import styled, { css } from 'styled-components';
import { Color } from '../config';

interface WrapperProps {
    pseudo: boolean
}

export const Wrapper = styled.span<WrapperProps>`
    border-bottom: ${props => (props.pseudo ? '1px dashed currentColor' : 'inherit')};
`;

interface StyledLinkProps {
    pseudo: boolean
}

export const StyledLink = styled.a<StyledLinkProps>`
    position: relative;
    display: inline;
    font-family: "Open Sans", Helvetica, Arial, sans-serif;
    font-size: inherit;
    font-weight: inherit;
    color: ${Color.Blue};
    cursor: pointer;
    text-decoration: none;
    &:hover {
        color: ${Color.DeepBlue};
        text-decoration: ${props => (props.pseudo ? 'none' : 'underline')};
    }
    &:hover ${Wrapper} {
        border-bottom-width: 0;
    }
`;

const getGutters = (props) => {
    let rightGutter = 0;
    let leftGutter = 0;
    if (props.pseudo) {
        rightGutter = 12;
        leftGutter = 12;
    }
    if (props.pseudo) {
        rightGutter = 4;
        leftGutter = 4;
    }
    return css`
        margin-right: ${rightGutter}px;
        margin-left: ${leftGutter}px;
    `;
};

export const Addon = styled.span`
    display: inline-block;
    width: 16px;
    height: 16px;
    vertical-align: middle;
    ${getGutters}
`;
