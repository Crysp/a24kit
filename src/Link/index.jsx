import React from 'react';
import PropTypes from 'prop-types';
import {
    StyledLink,
    Wrapper,
    Addon,
} from './styled';


/**
 * @develop
 */
export default class Link extends React.PureComponent {
    static propTypes = {
        /**
         * Текст ссылки
         */
        children: PropTypes.node.isRequired,
        /**
         * URL ссылки
         */
        href: PropTypes.string,
        /**
         * Иконка
         */
        icon: PropTypes.node,
        /**
         * Позиция иконки
         */
        iconPosition: PropTypes.oneOf(['before', 'after']),
        /**
         * Псевдо-ссылка
         */
        pseudo: PropTypes.bool,
        /**
         * Дополнительный CSS класс
         */
        className: PropTypes.string,
        /**
         * Обработчик клика
         */
        onClick: PropTypes.func,
    };

    static defaultProps = {
        href: '',
        iconPosition: 'before',
        pseudo: false,
        className: '',
        icon: null,
        onClick: undefined,
    };

    onClick = (e) => {
        const { href, onClick } = this.props;
        if (href.length === 0) e.preventDefault();
        if (onClick) onClick(e);
    };

    render() {
        const {
            children,
            href,
            icon,
            iconPosition,
            pseudo,
            ...other
        } = this.props;
        return (
            <StyledLink
                href={href}
                pseudo={pseudo}
                {...other}
                onClick={this.onClick}
            >
                <Wrapper pseudo={pseudo}>
                    {icon && iconPosition === 'before' && <Addon before>{icon}</Addon>}
                    <span>{children}</span>
                    {icon && iconPosition === 'after' && <Addon after>{icon}</Addon>}
                </Wrapper>
            </StyledLink>
        );
    }
}
