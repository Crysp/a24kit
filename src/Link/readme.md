```js
<Link>Ссылка</Link>
```

#### Ссылка с иконкой
```js
const { Chat } = require('../Icon');
<Link icon={<Chat/>}>Ссылка с иконкой</Link>
```

#### Псевдо-ссылка
```js
const { ArrowDown } = require('../Icon');
<div>
    <p><Link pseudo>Псевдо-ссылка</Link></p>
    <p><Link icon={<ArrowDown/>} pseudo>Псевдо-ссылка с иконкой</Link></p>
</div>
```