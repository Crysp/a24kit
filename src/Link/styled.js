import styled from 'styled-components';
import { Color } from '../config';

export const Wrapper = styled.span`
    border-bottom: ${props => (props.pseudo ? '1px dashed currentColor' : 'inherit')};
`;

export const StyledLink = styled.a`
    position: relative;
    display: inline;
    font-family: "Open Sans", Helvetica, Arial, sans-serif;
    font-size: inherit;
    font-weight: inherit;
    color: ${Color.Blue};
    cursor: pointer;
    text-decoration: none;
    &:hover {
        color: ${Color.DeepBlue};
        text-decoration: ${props => (props.pseudo ? 'none' : 'underline')};
    }
    &:hover ${Wrapper} {
        border-bottom-width: 0;
    }
`;

export const Addon = styled.span.attrs((props) => {
    let rightGutter = 0;
    let leftGutter = 0;
    if (props.pseudo) {
        rightGutter = 12;
        leftGutter = 12;
    }
    if (props.pseudo) {
        rightGutter = 4;
        leftGutter = 4;
    }
    return {
        rightGutter,
        leftGutter,
    };
})`
    display: inline-block;
    width: 16px;
    height: 16px;
    vertical-align: middle;
    margin-right: ${props => props.rightGutter}px;
    margin-left: ${props => props.leftGutter}px;
`;
