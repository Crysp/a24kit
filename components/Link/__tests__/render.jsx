import React from 'react';
import renderer from 'react-test-renderer';
import Link from '../index';


/* eslint-disable jsx-a11y/anchor-is-valid */
describe('Верстка компонента <Link>', () => {
    it('Дефолтная', () => {
        const tree = renderer.create(<Link href="#">Link</Link>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Дефолтная с иконкой', () => {
        const tree = renderer.create(<Link href="#" icon={<div>Icon</div>}>Link</Link>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Псевдоссылка', () => {
        const tree = renderer.create(<Link href="#" pseudo>Link</Link>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('С иконкой слева', () => {
        const tree = renderer.create(<Link href="#" icon={<div>Icon</div>} iconPosition="before">Link</Link>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('С иконкой справа', () => {
        const tree = renderer.create(<Link href="#" icon={<div>Icon</div>} iconPosition="after">Link</Link>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Дополнительний CSS класс', () => {
        const tree = renderer.create(<Link href="#" className="custom">Link</Link>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Прокидываются неописанные свойства', () => {
        const tree = renderer.create(<Link href="#" style={{ margin: 16 }}>Link</Link>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
/* eslint-enable jsx-a11y/anchor-is-valid */
