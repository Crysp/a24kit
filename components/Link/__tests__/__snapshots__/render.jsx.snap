// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Верстка компонента <Link> Дефолтная 1`] = `
<a
  className="sc-bwzfXH itTcug"
  href="#"
  onClick={[Function]}
>
  <span
    className="sc-bdVaJa bwiARo"
  >
    <span>
      Link
    </span>
  </span>
</a>
`;

exports[`Верстка компонента <Link> Дефолтная с иконкой 1`] = `
<a
  className="sc-bwzfXH itTcug"
  href="#"
  onClick={[Function]}
>
  <span
    className="sc-bdVaJa bwiARo"
  >
    <span
      className="sc-htpNat flmFzH"
    >
      <div>
        Icon
      </div>
    </span>
    <span>
      Link
    </span>
  </span>
</a>
`;

exports[`Верстка компонента <Link> Дополнительний CSS класс 1`] = `
<a
  className="custom sc-bwzfXH itTcug"
  href="#"
  onClick={[Function]}
>
  <span
    className="sc-bdVaJa bwiARo"
  >
    <span>
      Link
    </span>
  </span>
</a>
`;

exports[`Верстка компонента <Link> Прокидываются неописанные свойства 1`] = `
<a
  className="sc-bwzfXH itTcug"
  href="#"
  onClick={[Function]}
  style={
    Object {
      "margin": 16,
    }
  }
>
  <span
    className="sc-bdVaJa bwiARo"
  >
    <span>
      Link
    </span>
  </span>
</a>
`;

exports[`Верстка компонента <Link> Псевдоссылка 1`] = `
<a
  className="sc-bwzfXH eprhRL"
  href="#"
  onClick={[Function]}
>
  <span
    className="sc-bdVaJa dwEJpK"
  >
    <span>
      Link
    </span>
  </span>
</a>
`;

exports[`Верстка компонента <Link> С иконкой слева 1`] = `
<a
  className="sc-bwzfXH itTcug"
  href="#"
  onClick={[Function]}
>
  <span
    className="sc-bdVaJa bwiARo"
  >
    <span
      className="sc-htpNat flmFzH"
    >
      <div>
        Icon
      </div>
    </span>
    <span>
      Link
    </span>
  </span>
</a>
`;

exports[`Верстка компонента <Link> С иконкой справа 1`] = `
<a
  className="sc-bwzfXH itTcug"
  href="#"
  onClick={[Function]}
>
  <span
    className="sc-bdVaJa bwiARo"
  >
    <span>
      Link
    </span>
    <span
      className="sc-htpNat flmFzH"
    >
      <div>
        Icon
      </div>
    </span>
  </span>
</a>
`;
